### DISCUSSION TITLE
Writing a Webpack plugin with support for persistent caching

### DISCUSSION THREAD

**Discussion by privatenumber** [webpack/webpack#14270] [Q&A]
I'm looking to store plugin processing data into the persistent caching. Is there a property I can add serializable data to that I can re-read on cache-hit?

----

I'm developing a Webpack plugin & loader combo that optimizes/tree-shakes a JSON module given that it's only accessed using a specific getter function.

For example, given a build that contains one function call `getProperty('property-name')`, the plugin will strip all properties in the JSON module except `property-name`. (I believe Terser can handle something like this but this is a boiled down version of a more complex system that Terser could not optimize)

The plugin + loader has two steps:
1. Detects the usage of an arbitrary function (eg. `getProperty('property-name')`) via `normalModuleFactory.hooks.parser`
2. Removes unaccessed property names from a JSON module in the `compilation.hooks.finishModules` hook that has the loader applied to it, and runs `compilation.rebuildModule` for the loader to optimize the JSON module based on data from Step 1

This works fine on first build, but, when persistent cache is enabled, Step 1 (JavaScript parsing in modules) seems to be only executed for changed files. So the JSON optimization thinks only properties in the changed files are actually used.

How can I add the result from Step 1 into the persistent cache so that it doesn't need all JS to be re-parsed on a cache hit build?

**Comment by privatenumber** [comment]
Seems I'm able to store arbitrary data to `module.buildInfo` and expect to have it restored on cache hit.

As a follow-up, I'm curious if there's a hook to reinitialize state from module cache (prior to parsing non-cached modules).

Tapping into `normalModuleFactory.hooks.module` seems too early, where `buildInfo` isn't set yet.

**Reply by alexander-akait** [reply]
> Seems I'm able to store arbitrary data to module.buildInfo and expect to have it restored on cache hit.

yes, it is safe to store data in `buildInfo`

> As a follow-up, I'm curious if there's a hook to reinitialize state from module cache (prior to parsing non-cached modules).

In theory you can use cache hooks, but they will be design for other things
